weld/transforms/loop_fusion_2.rs

```diff
@@ -0,0 +1,524 @@
+//! Richer loop fusion rules.
+
+use ast::*;
+use ast::ExprKind::*;
+use ast::Type::*;
+use ast::BuilderKind::*;
+use error::*;
+
+use super::uniquify;
+
+use util::SymbolGenerator;
+
+extern crate fnv;
+
+struct MergeSingle<'a> {
+    params: &'a Vec<TypedParameter>,
+    value: &'a TypedExpr
+}
+
+impl<'a> MergeSingle<'a> {
+    fn extract(expr: &'a TypedExpr) -> Option<MergeSingle<'a>> {
+        if let Lambda{ref params, ref body} = expr.kind {
+            if let Merge{ref builder, ref value} = body.kind {
+                match builder.kind {
+                    Ident(ref name) if *name == params[0].name =>
+                        return Some(MergeSingle{params, value}),
+                    _ => {}
+                }
+            }
+        }
+        return None
+    }
+}
+
+struct NewAppender;
+
+impl NewAppender {
+    fn extract(expr: &TypedExpr) -> Option<NewAppender> {
+        if let NewBuilder(_) = expr.kind {
+            if let Builder(Appender(_), _) = expr.ty {
+                return Some(NewAppender)
+            }
+        }
+        return None
+    }
+}
+
+struct ResForAppender<'a> {
+    iters: &'a Vec<Iter<Type>>,
+    func: &'a TypedExpr
+}
+
+impl<'a> ResForAppender<'a> {
+    fn extract(expr: &'a TypedExpr) -> Option<ResForAppender<'a>> {
+        if let Res{ref builder} = expr.kind {
+            if let For{ref iters, ref builder, ref func} = builder.kind {
+                if NewAppender::extract(builder).is_some() {
+                    return Some(ResForAppender{iters, func})
+                }
+            }
+        }
+        return None;
+    }
+}
+
+struct MapIter<'a> {
+    iters: &'a Vec<Iter<Type>>,
+    merge_params: &'a Vec<TypedParameter>,
+    merge_value: &'a TypedExpr
+}
+
+impl<'a> MapIter<'a> {
+    fn extract(iter: &'a TypedIter) -> Option<MapIter> {
+        if iter.is_simple() {
+            if let Some(rfa) = ResForAppender::extract(&iter.data) {
+                if rfa.iters.iter().all(|ref i| i.is_simple()) {
+                    if let Some(merge) = MergeSingle::extract(&rfa.func) {
+                        return Some(MapIter {
+                            iters: &rfa.iters,
+                            merge_params: merge.params,
+                            merge_value: merge.value
+                        });
+                    }
+                }
+            }
+        }
+        return None;
+    }
+}
+
+/// Gets rid of `result(for(d, appender, _))` expressions within the iterators in a loop, replacing them
+/// with direct iteration over the data (e.g. `d` here).
+///
```

Comment:
Can you add an example of the transformation here? Also a better name would be good :P